phat/contracts/lego_rs/lib.rs

```diff
@@ -0,0 +1,152 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+#[macro_use]
+extern crate alloc;
+
+#[ink::contract]
+mod lego {
+    use crate::raw_bytes::RawBytes;
+    use alloc::string::String;
+    use alloc::vec::Vec;
+    use ink::env::Error;
+    use pink::{info, warn};
+    use scale::Encode;
+    use serde::Deserialize;
+    use this_crate::{version_tuple, VersionTuple};
+
+    type Result<T, E = Error> = core::result::Result<T, E>;
+
+    #[derive(Debug, Deserialize)]
+    #[serde(tag = "cmd", content = "config")]
+    #[serde(rename_all = "camelCase")]
+    enum Action {
+        Call {
+            #[serde(with = "deserialize_hex")]
+            callee: [u8; 32],
+            selector: u32,
+        },
+        Log,
+    }
+
+    mod deserialize_hex {
+        use super::*;
+
+        pub(super) fn deserialize<'de, D>(deserializer: D) -> Result<[u8; 32], D::Error>
+        where
+            D: serde::Deserializer<'de>,
+        {
+            let s = String::deserialize(deserializer)?;
+            let mut bytes = [0u8; 32];
+            hex::decode_to_slice(s.as_str().trim_start_matches("0x"), &mut bytes)
+                .map_err(serde::de::Error::custom)?;
+            Ok(bytes)
+        }
+    }
+
+    #[ink(storage)]
+    pub struct Lego {}
+
+    impl Lego {
+        #[ink(constructor)]
+        pub fn default() -> Self {
+            Self {}
+        }
+
+        #[ink(message)]
+        pub fn version(&self) -> VersionTuple {
+            version_tuple!()
+        }
+
+        #[ink(message)]
+        pub fn run(&self, actions: String) -> bool {
+            info!("lego_rs: actions={}", actions);
+            let Ok(actions) = pink_json::from_str::<Vec<Action>>(&actions) else {
+                warn!("failed to parse actions");
+                return false;
+            };
+            pipeline(&actions)
+        }
+    }
+
+    fn pipeline(actions: &[Action]) -> bool {
+        let mut result = String::new();
+        for action in actions {
+            info!("running action={:?}", action);
+            match action {
+                Action::Call { callee, selector } => {
+                    let res = invoke_contract(From::from(*callee), 0, 0, *selector, &[], false);
```

Comment:
How to pass the output from previous action to the latter one?

Reply:
Also it's okey to forbid reentry here, but we should record this in doc.

Reply:
> How to pass the output from previous action to the latter one?

No such feature at the moment.

Reply:
> Also it's okey to forbid reentry here, but we should record this in doc.

A full implementation will let the caller config the reentrant.